### PR TITLE
Incentive refactor: hard rewards

### PR DESCRIPTION
**Pull request by rhuairahrighairidh**
First of three PRs (#929, #928, #927) to add tests and simplify the incentive code.

This PR refactors and tests the hard borrow and supply rewards.
The behaviour is (mostly) unchanged.
Some TODOs have been cleaned up in the #927

There's loads of commits, but everything up to `Merge branch 'master' into ro-incentive-refactor-rebase-temp` is already in master.

changes

- Add a unit tests for Initialize..., Synchronize..., Update... methods
  - aims:
    - ensure I don't break anything by specifing what the method should do
    - try out a structure for unit tests with no dependencies
      - removing dependencies should simplify the test setup
- Refactor of Initialize..., Synchronize..., Update... methods
  - aims:
    - move (Multi)RewardIndex manipulations out of these functions to make them a little easier to read
    - rearrange these functions to look the same, to help with deduplication later
  - changes
    - Added With and Get methods to (Multi)RewardIndexes to encapsulate read/writes
    - If a global reward factor ( ie "bnb": "hard": 0.12345) somehow dissapears Sync will now panic. Before there was no panic, and the old factor was left in the user indexes.

### CODE REVIEW

**Review by karzak — CHANGES_REQUESTED** [review]
Looks good - mostly minor stuff and clarifications.

File: x/incentive/types/claims.go [hunk 1]
```diff
@@ -514,3 +563,10 @@ func (mris MultiRewardIndexes) Validate() error {
 	}
 	return nil
 }
+
+// copy returns a copy the slice and underlying array
```
[the remark this answers is not in the document]
Comment (rhuairahrighairidh) [reply x/incentive/types/claims.go:567]: fixed in #927

File: x/incentive/keeper/rewards_borrow.go [hunk 2]
```diff
@@ -113,44 +111,20 @@ func (k Keeper) SynchronizeHardBorrowReward(ctx sdk.Context, borrow hardtypes.Bo
 			continue
 		}
 
-		userMultiRewardIndex, foundUserMultiRewardIndex := claim.BorrowRewardIndexes.GetRewardIndex(coin.Denom)
-		if !foundUserMultiRewardIndex {
+		userRewardIndexes, found := claim.BorrowRewardIndexes.Get(coin.Denom)
+		if !found {
 			continue
```
[the remark this answers is not in the document]
Comment (rhuairahrighairidh) [reply x/incentive/keeper/rewards_borrow.go:116]: good point

File: x/incentive/keeper/rewards_borrow.go [hunk 3]
```diff
@@ -165,30 +139,74 @@ func (k Keeper) UpdateHardBorrowIndexDenoms(ctx sdk.Context, borrow hardtypes.Bo
...
 	k.SetHardLiquidityProviderClaim(ctx, claim)
 }
+
+// CalculateRewards computes how much rewards should have accrued to a source (eg user's hard borrowed btc amount)
+// between two index values.
+//
+// oldIndex is normally the index stored on a claim, newIndex the current global value, and rewardSource a hard borrowed/supplied amount.
+func (k Keeper) CalculateRewards(oldIndexes, newIndexes types.RewardIndexes, rewardSource sdk.Int) (sdk.Coins, error) {
+	for _, oldIndex := range oldIndexes {
+		if newIndex, f := newIndexes.Get(oldIndex.CollateralType); !f {
```
[the remark this answers is not in the document]
Comment (rhuairahrighairidh) [reply x/incentive/keeper/rewards_borrow.go:172]: fixed in #927

File: x/incentive/keeper/rewards_supply.go [hunk 4]
```diff
@@ -116,45 +112,20 @@ func (k Keeper) SynchronizeHardSupplyReward(ctx sdk.Context, deposit hardtypes.D
 			continue
 		}
 
-		userMultiRewardIndex, foundUserMultiRewardIndex := claim.SupplyRewardIndexes.GetRewardIndex(coin.Denom)
+		userMultiRewardIndexes, foundUserMultiRewardIndex := claim.SupplyRewardIndexes.Get(coin.Denom)
```
[the remark this answers is not in the document]
Comment (rhuairahrighairidh) [reply x/incentive/keeper/rewards_supply.go:115]: fixed in #929, I standardized a few other variables names as well

File: x/incentive/keeper/rewards_borrow.go [hunk 5]
```diff
@@ -165,30 +139,74 @@ func (k Keeper) UpdateHardBorrowIndexDenoms(ctx sdk.Context, borrow hardtypes.Bo
...
+		)
+	}
+	return reward, nil
+}
+
+// calculateSingleReward computes the rewards that should accumulate between two index values.
+
+// oldIndex is normally the index stored on a claim, newIndex the current global value, and rewardSource a hard borrowed/supplied amount.
+// newIndex MUST be greater than oldIndex otherwise it will panic
+func (k Keeper) calculateSingleReward(oldIndex, newIndex sdk.Dec, rewardSource sdk.Int) sdk.Int {
```
[the remark this answers is not in the document]
Comment (rhuairahrighairidh) [reply x/incentive/keeper/rewards_borrow.go:205]: agreed, that's a nicer approach. Fixed in #927, https://github.com/Kava-Labs/kava/pull/927/commits/c5d31948cefb4cbb3e2ec830a905bac912fc83a3

File: x/incentive/keeper/unit_test.go [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+	suite.keeper = suite.setupKeeper(incentiveStoreKey)
+	suite.ctx = NewTestContext(incentiveStoreKey)
+}
+
+func (suite *unitTester) TearDownTest() {
+	suite.keeper = keeper.Keeper{}
+	suite.ctx = sdk.Context{}
+}
+
+func (suite *unitTester) setupKeeper(incentiveStoreKey sdk.StoreKey) keeper.Keeper {
```
[the remark this answers is not in the document]
Comment (rhuairahrighairidh) [reply x/incentive/keeper/unit_test.go:53]: Not using TestApp was a bit of an experiment. I wanted to try setting up unit tests without importing anything from other modules or `app/`. Then any changes outside this module can't break the tests.

I think we still need `TestApp` tests to make sure the modules work together.

I was thinking that if we had unit tests covering each keeper method, it'd free us up to make our `TestApp` tests more realistic by matching what happens on mainnet, ie by delivering messages, running full end blockers, etc. Bit like the delegation syncing tests at the end of https://github.com/Kava-Labs/kava/blob/ro-incentive-refactor-tidy-hard/x/incentive/keeper/rewards_delegator_test.go

Keen to get feedback on this approach

**Pull request merged**